File: android/farmerchat-sdk/src/main/java/com/farmerchat/sdk/ui/chat/udf/ChatAction.kt
package com.farmerchat.sdk.ui.chat.udf

import android.net.Uri

sealed class ChatAction {

    data class InitializeWithPreGeneratedContent(
        val question: String,
        val answer: String?,
        val followUpQuestions: List<String>?
    ) : ChatAction()

    data class InitializeWithQuestion(
        val question: String,
        val transcriptionId: String? = null,
        val audioUri: Uri? = null
    ) : ChatAction()

    data class ReplacePreGeneratedWithQuestion(
        val question: String,
        val triggerInputType: String? = null
    ) : ChatAction()

    data class SendFollowUpQuestion(
        val question: String,
        val followUpQuestionId: String? = null,
        val transcriptionId: String? = null,
        val audioUri: Uri? = null
    ) : ChatAction()

    data class SendQuestionWithImage(
        val question: String,
        val imageUri: Uri
    ) : ChatAction()

    data class SendQuestionWithAudio(
        val question: String,
        val audioUri: Uri
    ) : ChatAction()

    /** Raw audio bytes (base64) → transcribe API → then send as text query */
    data class TranscribeAndSendAudio(
        val audioBase64: String,
        val audioFormat: String,
        val audioUri: Uri?
    ) : ChatAction()

    data class LoadChatHistory(
        val conversationId: String,
        val page: Int = 1
    ) : ChatAction()

    object RetryLastRequest : ChatAction()
    object ClearError : ChatAction()
    object ClearMessages : ChatAction()
    object SynthesiseAudio : ChatAction()
    object ClearAudioPlaybackUrl : ChatAction()

    data class SetAudioPlaying(val isPlaying: Boolean) : ChatAction()
}

File: ios/Sources/FarmerChatSDK/Models/UI/ChatAction.swift
import Foundation

// MARK: - ChatAction

/// All events that can be dispatched to `ChatViewModel`.
public enum ChatAction {
    /// User sends their first question in a new conversation.
    case initializeWithQuestion(
        question: String,
        transcriptionId: String?,
        audioUrl: URL?
    )

    /// The conversation starts with pre-generated content (no API call for the first turn).
    case initializeWithPreGeneratedContent(
        question: String,
        answer: String?,
        followUpQuestions: [String]?
    )

    /// Replace a pre-generated message pair with a live API call.
    case replacePreGeneratedWithQuestion(
        question: String,
        triggerInputType: String?
    )

    /// User taps a follow-up suggestion chip.
    case sendFollowUpQuestion(
        question: String,
        followUpQuestionId: String?,
        transcriptionId: String?,
        audioUrl: URL?
    )

    /// User sends a question with an attached image.
    case sendQuestionWithImage(question: String, imageData: Data)

    /// User sends a voice message (audio file on-device).
    case sendQuestionWithAudio(question: String, audioUrl: URL)

    /// Raw audio data → transcribe API → send as text query.
    case transcribeAndSendAudio(audioData: Data, audioUrl: URL?)

    /// Load paginated chat history for a given conversation.
    case loadChatHistory(conversationId: String, page: Int)

    /// Retry the last failed API action.
    case retryLastRequest

    /// Dismiss the current inline error without retrying.
    case clearError

    /// Clear all messages and reset to empty state.
    case clearMessages

    /// Request server-side text-to-speech for the latest AI message.
    case synthesiseAudio

    /// Clear the audio playback URL (e.g. after playback finishes).
    case clearAudioPlaybackUrl

    /// Update the audio playing flag.
    case setAudioPlaying(Bool)
}
